Factory / Singleton constructor parameters typing

Currently, the type of the  `factory` constructor parameter of `Factory` / `Singleton` providers is defined as a union `type[T_co] | typing.Callable[P, T_co]`. 

Because of this, Mypy doesn't detect issues with invalid arguments for the constructor. For example:

```Python
@dataclasses.dataclass
class Dependency:
    foo: str
    bar: int


class DIContainer(BaseContainer):
    factory_1 = providers.Factory(Dependency)
    factory_2 = providers.Factory(Dependency, invalid_foo="foo", invalid_bar="bar")
    singleton_1 = providers.Singleton(Dependency)
    singleton_2 = providers.Singleton(Dependency, invalid_foo="foo", invalid_bar="bar")
```

Mypy shows **no errors** in this case.

If we change the type annotation of `factory` just to `typing.Callable[P, T_co]`, then Mypy starts to show the expected errors:
```
tests/providers/test_temp.py:12: error: Missing positional arguments "foo", "bar" in call to "Factory"  [call-arg]
tests/providers/test_temp.py:13: error: Unexpected keyword argument "invalid_foo" for "Factory"  [call-arg]
that_depends/providers/factories.py:13: note: "Factory" defined here
tests/providers/test_temp.py:13: error: Unexpected keyword argument "invalid_bar" for "Factory"  [call-arg]
that_depends/providers/factories.py:13: note: "Factory" defined here
tests/providers/test_temp.py:14: error: Missing positional arguments "foo", "bar" in call to "Singleton"  [call-arg]
tests/providers/test_temp.py:15: error: Unexpected keyword argument "invalid_foo" for "Singleton"  [call-arg]
that_depends/providers/singleton.py:14: note: "Singleton" defined here
tests/providers/test_temp.py:15: error: Unexpected keyword argument "invalid_bar" for "Singleton"  [call-arg]
that_depends/providers/singleton.py:14: note: "Singleton" defined here
```

Since `type[T]` is also a callable in Python, I'd suggest updating the type annotation to improve Mypy type checking. What do you think?

--- a/tests/tricky_cases/test_falsy_instances.py
+++ b/tests/tricky_cases/test_falsy_instances.py
@@ -20,7 +20,7 @@ async def create_async_resource() -> typing.AsyncIterator[dict[str, typing.Any]]
 
 
 class TrickyDIContainer(BaseContainer):
-    singleton = providers.Singleton(list)
+    singleton = providers.Singleton(list[object])
     sync_resource = providers.Resource(create_sync_resource)
     async_resource = providers.Resource(create_async_resource)
 

--- a/that_depends/providers/factories.py
+++ b/that_depends/providers/factories.py
@@ -10,7 +10,7 @@ P = typing.ParamSpec("P")
 class Factory(AbstractFactory[T_co]):
     __slots__ = "_factory", "_args", "_kwargs", "_override"
 
-    def __init__(self, factory: type[T_co] | typing.Callable[P, T_co], *args: P.args, **kwargs: P.kwargs) -> None:
+    def __init__(self, factory: typing.Callable[P, T_co], *args: P.args, **kwargs: P.kwargs) -> None:
         super().__init__()
         self._factory: typing.Final = factory
         self._args: typing.Final = args

--- a/that_depends/providers/singleton.py
+++ b/that_depends/providers/singleton.py
@@ -11,7 +11,7 @@ P = typing.ParamSpec("P")
 class Singleton(AbstractProvider[T_co]):
     __slots__ = "_factory", "_args", "_kwargs", "_override", "_instance", "_resolving_lock"
 
-    def __init__(self, factory: type[T_co] | typing.Callable[P, T_co], *args: P.args, **kwargs: P.kwargs) -> None:
+    def __init__(self, factory: typing.Callable[P, T_co], *args: P.args, **kwargs: P.kwargs) -> None:
         super().__init__()
         self._factory: typing.Final = factory
         self._args: typing.Final = args

